### PR TITLE
Gun Overheat Mechanics

### PR DESCRIPTION
**Pull request by tyeagg**
## About The Pull Request
not actually used on any gun right now but for future pr.
adds overheat mechanics to guns if they use the overheat mechanics var when overheat cap is reached gun will be forced to stop firing for whatever set amount of seconds to cool down

example:

https://user-images.githubusercontent.com/98577688/201936758-53a1b439-2cb1-481b-a9ec-5b8df35e63ba.mp4
## Why It's Good For The Game
cool overheat mechanics that could be used for cool guns and balancing
## Changelog
:cl:
add: overheat mechanics to guns if they use them
/:cl:

### CODE REVIEW

**Comment by FIXTHERADBOXPLEASE** [comment]
Instamerge and add to both miniguns.

**Comment by XSlayer300** [comment]
Gotta have a overheat warning by the white smoke appearing on the sprite by 60% heat.

**Comment by Losenis** [comment]
I am extremely hopeful that this leads to laser guns that actaully have a passive recharge rather than reload mechanic

**Comment by Skatuk** [comment]
Pretty cool, would be a good balance to the RPM

**Comment by Losenis** [comment]
Also this does mean there could be an attachment in the future that gives faster firerate in exchange for overheating

**Comment by FosterGrime-Heymynameslouie** [comment]
> I am extremely hopeful that this leads to laser guns that actaully have a passive recharge rather than reload mechanic

yes plea

also spedemgher

**Comment by Starfirejordan** [comment]
Good shit

**Comment by Al-1ce** [comment]
It's a nice idea but I'm skeptical to see this work in actual gameplay outside of very seldom examples. Probably too harsh of a nerf for SGs since their guns don't do a whole lot of damage on their own and primarily rely on their teammates.

File: code/modules/projectiles/gun_system.dm [hunk 1]
```diff
@@ -455,6 +481,17 @@
 		for(var/datum/action/action AS in master_gun.actions)
 			action.update_button_icon()
 
+	if(use_heat)
+		if(overheated)
+			if(overheat_smoke)
+				return
+			overlays += heat_overlay
```
Comment (Lumipharon) [inline code/modules/projectiles/gun_system.dm:488]: Overlays should be handled in update_overlays()

**Review by Lumipharon — CHANGES_REQUESTED** [review]
Cool mechanic, the amount of smoke is kinda funny though.

File: code/modules/projectiles/gun_system.dm [hunk 2]
```diff
@@ -455,6 +481,17 @@
 		for(var/datum/action/action AS in master_gun.actions)
 			action.update_button_icon()
 
+	if(use_heat)
+		if(overheated)
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:485]: You can combine checks

File: code/modules/projectiles/gun_system.dm [hunk 3]
```diff
@@ -455,6 +481,17 @@
...
+			if(overheat_smoke)
+				return
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:487]: so if there is smoke you do no update item state ?

File: code/modules/projectiles/gun_system.dm [hunk 4]
```diff
@@ -1534,6 +1574,12 @@
 	if(CHECK_BITFIELD(flags_gun_features, GUN_IS_ATTACHMENT) && !master_gun && CHECK_BITFIELD(flags_gun_features, GUN_ATTACHMENT_FIRE_ONLY))
 		to_chat(user, span_notice("You cannot fire [src] without it attached to a gun!"))
 		return FALSE
+	if(heat_level >= max_heat || overheated)
+		user.balloon_alert(user, "The gun is overheated!")
+		if(!overheat_timer)
+			overheated = TRUE
+			overheat_timer = addtimer(CALLBACK(src, .proc/cooldown), cooldown_time, TIMER_STOPPABLE)
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:1581]: so you start the timer when you are trying to shoot ?

File: code/modules/projectiles/gun_system.dm [hunk 5]
```diff
@@ -1534,6 +1574,12 @@
 	if(CHECK_BITFIELD(flags_gun_features, GUN_IS_ATTACHMENT) && !master_gun && CHECK_BITFIELD(flags_gun_features, GUN_ATTACHMENT_FIRE_ONLY))
 		to_chat(user, span_notice("You cannot fire [src] without it attached to a gun!"))
 		return FALSE
+	if(heat_level >= max_heat || overheated)
```
Comment (BraveMole) [inline code/modules/projectiles/gun_system.dm:1577]: why do you have overheated and max_heat

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/gun_system.dm (hunk 4 above)
Comment (tyeagg) [reply code/modules/projectiles/gun_system.dm:1581]: yes

Comment (BraveMole) [reply code/modules/projectiles/gun_system.dm:1581]: That seems really frustrating

**Pull request closed without merging**